### ISSUE TITLE
Consolidate resumption PSK definitions

### ISSUE THREAD

**Issue by bifurcation** [mlswg/mls-protocol#502]
Right now, we treat the re-init and branch PSK cases as distinct cases, as opposed to just having a general "export PSK from one group+epoch, insert it into another" function.  What I would call a "resumption PSK", following TLS.   Even though we have the distinction, the content of the PSK itself is the same: (group ID, epoch, resumption secret).  So the only difference is that the PreSharedKeyID has a bit that signals the intent of injecting this resumption PSK.

Do we really need that bit?  If there's not a compelling need, it would streamline the API and allow for more flexibility if we just had the notion of a resumption PSK.

**Comment by kkohbrok** [comment]
Key separation by domain is usually a good idea. If we derive a key, the purpose should be "baked in", i.e. included in the derivation to avoid potential collision across protocol functionalities/domains. I'm not sure I understand how that makes the API more complex. If it's really bad, we should consider the trade-off, but if it's not too much trouble, I'd prefer we keep the key separation. Can you elaborate a bit on the API complexity?

**Comment by bifurcation** [comment]
I'm mostly concerned about being too specific in the "purpose" definitions, and thus making applications make a choice that doesn't really make sense.  For example, take the obvious case of TLS-like resumption, where I take a key from the last epoch of a prior instance of the group and use it as input to a new appearance of the group.  Is that a re-init?  There wasn't necessarily a ReInit proposal.  Is it a branch?  "Branch" isn't even defined in the spec.  Better to just talk about "resumption" as the general act of exporting a secret from one group and inserting it into another.

**Comment by kkohbrok** [comment]
The scenarios you describe seem to indicate problems in the spec that don't have anything to do with the derivation of the resumption secret. If you want to enable re-init secrets without a proposal, that's fine and we can define in the spec which key to use for that purpose (probably the re-init key). Similarly, if branch isn't well defined in  the spec, we should probably define it better. In the end, for everything else that you want to do that's not (properly) defined in the spec, you can just use an exporter secret. I don't see that weakening the domain separation of the re-init and branch key derivation is the right way to solve these problems. Is the code complexity that bad?

**Comment by bifurcation** [comment]
We could go off and try to define all the possible uses of a resumption PSK, but (a) that's a lot more work, (b) it seems unlikely to succeed, (c) even if you do succeed, you'll end up with a more complicated API for applications (bc N names for the same technical act) and (d) it's not clear to me what the benefit is.

When you talk about domain separation here, what is the failure case that this prevents?

**Comment by kkohbrok** [comment]
What we're trying to prevent is that participants end up with the same keys after executing a different part/functionality of the protocol. If party A thinks, they are doing a branch, while party B thinks they're doing a Re-Init and they end up with the same key, then that's bad. If an implementer wants to do proprietary things, i.e. things that are not described in the spec, they have to use an exporter key and make sure their keys are properly separated.

**Comment by bifurcation** [comment]
It depends on what you mean by a "different functionality of the protocol".  The only difference between the branch and reinit cases is that in the reinit case, you quit using the old group.  You could also imagine injecting a resumption PSK into an independent group with its own history, whose membership overlaps with the group from which the PSK was drawn.  Is the functionality "linking one group/epoch to a prior group/epoch" or is it "reinit / branch / ..."?  The former seems more at the right level of granularity to me, and leads to the approach here and in #518.

If you're really hard over here, how about a compromise: Unify the syntax, but add a `usage` distinguisher alongside the group ID and epoch, which has specified values for the branch and reinit cases, and which the application MAY set for other cases.

```
enum {
  unspecified(0),
  branch(1),
  reinit(2),
  (255)
} PSKUsage;

// ...

    case resumption:
      PSKUsage usage;
      opaque psk_group_id<0..255>;
      uint64 psk_epoch;
```

**Comment by kkohbrok** [comment]
I see where you're coming from and if you're convinced it is the more useful approach, we can make the trade-off and go with #518 (although we might as well rename the key from "resumption" to something like "group-link" or something). 

However, I do want to note that we have a trade-off here. Previously, designating a PSK "branch" or "re-init" signaled intent on the side of the sender and arriving at the same PSK meant agreement on said intent. For example, using a re-init key would imply that the new group is meant to replace the old group. As far as I understand that is no longer the case with #518.

**Comment by bifurcation** [comment]
Even if we want to have everyone agree that the group should be shut down, the ReInit proposal provides that.  Committing the ReInit puts it in the transcript, which means everyone agrees on it.  So everyone who exports a PSK from that epoch should know it's a "dead epoch".  In fact, the `branch` vs. `reinit` distinction seems duplicative here, since you shouldn't be able to get a `reinit` PSK from a "live" epoch or a `branch` PSK from a dead one.  `branch` iff live; `reinit` iff dead.

Now, there may be other sub-cases within live or dead; as noted, not all resumption PSK cases map clearly to `branch` or `reinit`.  But it seems premature to put in stub functionality without some taxonomy and requirements.

I would propose we do #518 for now, and if there turns out to be a need to be PSK use cases that really need segregating, it should be straightforward extension to add another PSKType.

**Comment by kkohbrok** [comment]
Fair enough. I see that the trade-off is probably ok. I'd be ok with merging #518.

The proposal should indeed ensure that the streams don't get crossed, so that all that remains would be ease of provability at that point. I still feel uncomfortable about not properly separating keys, though.

Regarding the last two arguments:
- the "for now" solution is a bit moot, since with the WGLC at hand, it's almost certainly not going to change until then
- the extension argument goes the other way as well, as you could just as easily have an extension that implements #518 and gets rid of the need for the distinction between resumption PSKs. If anything, we should opt to be conservative in the spec and allow users to loosen the key separation via an extension at their own risk.

**Comment by kkohbrok** [comment]
> I understand the concern from a cryptographic standpoint, it's nice to have a type that indicates that the group should be shut down. From an engineering perspective, this all feels pretty vague. It's not clear at all how exactly a group will be shut down and a new one will be created, the protocol doesn't give any specific guidance. I don't think this is a huge concern, as long as we don't paint ourselves in a corner when we want to upgrade groups to newer MLS versions in the future.

I understand that it's unclear how the linking of groups is going to take place in a practical setting, but I don't see that that's an argument to weaken the cryptographic basis of the mechanism. If we don't figure out how to do it in practice, then that's fine, as we can just ignore that part of the protocol. If we do, however, we will want to get proper security guarantees whenever we link two groups in this way.

Again, I think #518 is ok. But it's because we probably get good enough guarentees and not because the whole mechanism is a secondary concern due to there being unsolved higher level engineering challenges.

**Comment by raphaelrobert** [comment]
I have recently implemented some of this to get a better feel for it. While we already knew the "reinit" case was quite underspecified, I now also realized that there might be another issue with it: A member can shut down a group before a new one is created and there is no way to force the creation of the new group within the protocol. I think it would be safer to first create a new group (with whatever new parameters) and to tear down the old one in that process. Since this is too complex for the protocol it should be solved at the application layer. In other words, I think #518 is indeed fine.

**Comment by br-hale** [comment]
Commenting on #518 here, to keep discussion from splitting to the PR.

There are a few issues here that could use clarify, and some confusion in the interpretation of the proposals.

First - as @kkohbrok pointed out, if anything is unclear/underspecified, then that should be clarified and cleaned up vs. removal as the default. From the confusion above, it is clear that some pieces are not sufficiently specified.

Second - There is a core distinction between how PSKs are used in MLS vs. TLS that we need to be careful of. In TLS, these are used strictly for the same _two_ members communicating again after a finished session. It is not used again within the same session nor outside of the protocol (exporter keys are for that). Namely, the uses are:

1. Group-internal use (resumption by an identical member set of the two parties)
 
In MLS, these assumptions change by the very fact that the group may have more than 2 members. For the protocol-external case we have exporter keys as well. For PSK use, we have the following:

1. Group-internal use (resumption by an identical member set)
2. Subgroup-internal use (resumption by a proper subset of members)
3. Session-internal use (since continuous KE implies that keying material may be injected)

While the session-internal use is not the focus of this discussion, it has been previously (i.e. proving knowledge of prior group state). This contrasts against TLS were the handshake takes place only once

Point 2 is notable in that it is impossible for a proper subset of TLS session members to use session information in a separate session (i.e. that would imply only one communication party with no partner). In MLS, this is possible, and we need to be careful about mis-aligning TLS for precedent. To avoid potential forks, etc., and clarity of key source, any group-subset PSK use really ought to be denoted (currently a 'branch'). @kkohbrok is correct that there could be subtle attacks on this if key use is not properly separated.


In addition to the above there seems to be confusion on what resumption is linked to (this is consequently something that needs more clarity in the spec.). To be precise in alignment of terms, using a PSK in TLS for communication in a new session is aligned to using a PSK in MLS for a new group, i.e. it being a continuous session. Thus we need to align discussion here between "alive" and "dead" sessions in TLS to "alive" and "dead" groups in MLS - not epochs. We can get a reinit PSK from a live group, i.e. if a member wants to force a restart for any reason and instantiates a new group with the reinit secret. This signals to all members joining that group that the prior group should be terminated and not considered secure/alive after the point of reinit use. It is an edge case, but possible. Similarly and more commonly, we could get a branch PSK from a dead group, i.e. if a new (sub-)group wants members to prove knowledge from some prior connection. Thus it follows that we can branch from both "alive" and "dead" groups. This may be helpful in terms of authentication.

I am favorable to @bifurcation 's PSKUsage suggestion, i.e. that if there are other desired uses apart from the above cases the application can provide for those. The "bleeding" of key separation across different uses is something that is concerning, and should involve closer scrutiny before inclusion in the specification. If anything, more separation is safer at this stage, vs consolidation.

**Comment by bifurcation** [comment]
@br-hale Good point about the differences from the two-party case.  I think your taxonomy is not quite complete, though. You could also imagine cases where the epoch into which the resumption PSK is injected has members that were not in the group in the epoch in which it was extracted (and these members were provided the PSK some other way).  For instance, in the session-internal case where there have been some Adds in the meantime.

To try to factor this a little differently, it seems like you have at least three independent columns in the truth table here comparing the extraction epoch to the injection epoch:

* Same group or different group ("group" is the term that matches TLS "session"; linear sequence of epochs)
* Some members have been removed
* Some members have been added

So you have at least 8 cases to cover to start with assuming you care about both of those axes (same/different group and membership).

It's not clear to me why those are the two axes that one cares about.  On the one hand, why is it not important to distinguish ciphersuite changes as well, or changes in extensions?  On the other hand, why do we need to agree on membership changes here, given that the key schedule already confirms that we agree on the membership?  Basically, I don't understand (a) what cases need to be distinguished, (b) why do these cases need to be distinguished, and (c) why the required separation properties aren't provided by some other mechanism already.

I would also note that you don't get the benefits of key separation unless the recipients know how to verify that the usage is right for the injection context. (Otherwise, a malicious PSK sender can send the wrong value.)  For the same/different group, that's easy enough to do based on key schedule continuity.  But membership is problematic.  For example, if you're changing the ciphersuite between extraction and injection, then all of the KeyPackages and Credentials will be different -- what do you compare to verify equality?

I can update #518 to add PSKUsage, but the cases we put there need to at least be enforceable, and ideally fit into some principled taxonomy (even if all the cases in the taxonomy aren't enforceable).

**Comment by br-hale** [comment]
For the record, distinguishing on ciphersuite changes was brought up earlier in the WG - I am in the support if bringing it back on the discussion table is an option...
Separately though, for most other changes during the lifespan of a group, proposals basically perform a similar action for distinguishing between keys and separating changes from epoch to epoch. The difference with the PSK is there is no "one way" to interpret a PSK's use (as compared to e.g. a ciphersuite change). This also means that it can be more challenging to provably separate out misuse. 
[BTW: I am assuming in this discussion you mean ciphersuite change in the context of the listed components, since reinit is actually used for ciphersuite key changes per line 1598.]

You are quite correct that there is a group-external use, i.e. the "different group" case. However, anything outside of the group/session is an external/exporter key use. Namely, the MLS protocol spec is about a single session and anything that is used in other groups/sessions is an external key (similar to TLS export from session). The distinction between PSK usage and external/exporter keys (whether the latter is used for another application or injected into a different group entirely) is that PSK use is within the analysis bounds of the given session. To this point, there is a gray area on item 2 (subgroup-internal use / resumption by a proper subset of members). This is probably the only one I could see an argument for also pushing as an exporter key use. 

The issue of some members added between PSK usage is really reducible to two cases: a) a member was already in the set and the PSK use appears as item 1 or 3 (group-internal use or session-internal use) or b) a member was added and the PSK offers no security beyond that of an exporter key inject. The potential value proposition is for existing members, there is no loss for a new member, and the it is better to offer the value as possible than to deny it based on the weakest link.

The issue of some members being removed between PSK usage is similarly reducible. If the ejected member does not have the current group state then knowing a PSK should not provide value. It is similar to knowing an exporter key but having no place to export it into. The goal of PSK usage is to strengthen an existent system vs. being the sole point of security for that. If our KDF is done correctly, then combing current group state and the PSK will not be problematic (former unknown but latter known to the ejected member).

**Comment by bifurcation** [comment]
So @kkohbrok and I had some offline discussion about this, and it led me to the conclusion that what I'm really grumpy about is a need for more clarity in the branch and reinit definitions.  We don't need a full taxonomy of uses if (a) we can say "use this PSK usage for this specific protocol feature", and (b) we have some escape hatch that allows for usages of resumption PSKs outside those specific protocol features.

I have filed #525 reflecting this theory.  Hopefully folks find it more agreeable.

**Comment by br-hale** [comment]
That seems like a well-reasoned solution.

**Comment by bifurcation** [comment]
Fixed with #518  and #525
